// ng2-json-editor/src/bottom-console/patches-console-tab/patches-console-tab.component.js
/*
 * This file is part of ng2-json-editor.
 * Copyright (C) 2017 CERN.
 *
 * ng2-json-editor is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License as
 * published by the Free Software Foundation; either version 2 of the
 * License, or (at your option) any later version.
 *
 * ng2-json-editor is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with ng2-json-editor; if not, write to the Free Software Foundation, Inc.,
 * 59 Temple Place, Suite 330, Boston, MA 02111-1307, USA.
 * In applying this license, CERN does not
 * waive the privileges and immunities granted to it by virtue of its status
 * as an Intergovernmental Organization or submit itself to any jurisdiction.
 */
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
import { Component, ChangeDetectionStrategy, ChangeDetectorRef } from '@angular/core';
import { AbstractSubscriberComponent } from '../../abstract-subscriber';
import { DomUtilService, PathUtilService, JsonStoreService } from '../../shared/services';
var PatchesConsoleTabComponent = (function (_super) {
    __extends(PatchesConsoleTabComponent, _super);
    function PatchesConsoleTabComponent(domUtilService, pathUtilService, jsonStoreService, changeDetectorRef) {
        var _this = _super.call(this) || this;
        _this.domUtilService = domUtilService;
        _this.pathUtilService = pathUtilService;
        _this.jsonStoreService = jsonStoreService;
        _this.changeDetectorRef = changeDetectorRef;
        _this.patches = [];
        return _this;
    }
    PatchesConsoleTabComponent.prototype.ngOnInit = function () {
        var _this = this;
        this.jsonStoreService.patchesByPath$
            .map(function (patchesByPath) {
            return Object.keys(patchesByPath)
                .map(function (path) { return patchesByPath[path]; })
                .reduce(function (pre, cur) { return pre.concat(cur); }, []);
        })
            .takeUntil(this.isDestroyed)
            .subscribe(function (patches) {
            _this.patches = patches;
            _this.changeDetectorRef.markForCheck();
        });
    };
    PatchesConsoleTabComponent.prototype.focusPatch = function (patch) {
        this.domUtilService.focusPatch(patch);
    };
    PatchesConsoleTabComponent.prototype.acceptAll = function () {
        var _this = this;
        this.patches
            .forEach(function (patch) { return _this.jsonStoreService.applyPatch(patch); });
    };
    PatchesConsoleTabComponent.prototype.rejectAll = function () {
        var _this = this;
        this.patches
            .forEach(function (patch) { return _this.jsonStoreService.rejectPatch(patch); });
    };
    return PatchesConsoleTabComponent;
}(AbstractSubscriberComponent));
export { PatchesConsoleTabComponent };
PatchesConsoleTabComponent.decorators = [
    { type: Component, args: [{
                selector: 'patches-console-tab',
                styles: [".list-group-item { cursor: pointer; color: #0074D9; } .list-group-item:hover { background-color: #faebcc; text-decoration: underline; } ",
                    ".fa-bolt { color: #e67e22; } .icon-padding-left { padding-left: 8px; } .all-actions-container { padding: 8px; } .all-actions-container > button { margin-right: 4px; margin-left: 4px; } "],
                template: "<ng-template tabHeading> <i class=\"fa fa-bolt\"></i> Conflicts <span class=\"badge\">{{patches.length}}</span> </ng-template> <div *ngIf=\"patches && patches.length > 0\" class=\"all-actions-container\"> <button class=\"btn btn-success\" (click)=\"acceptAll()\">Accept All<i class=\"fa fa-check icon-padding-left\"></i></button> <button class=\"btn btn-danger\" (click)=\"rejectAll()\">Reject All<i class=\"fa fa-times icon-padding-left\"></i></button> </div> <ul class=\"list-group\"> <li *ngFor=\"let patch of patches\" class=\"list-group-item\" (click)=\"focusPatch(patch)\"> <i class=\"fa fa-bolt\"></i> {{patch.path}} - {{patch.op}} </li> </ul> ",
                changeDetection: ChangeDetectionStrategy.OnPush
            },] },
];
/** @nocollapse */
PatchesConsoleTabComponent.ctorParameters = function () { return [
    { type: DomUtilService, },
    { type: PathUtilService, },
    { type: JsonStoreService, },
    { type: ChangeDetectorRef, },
]; };
